### PR TITLE
[WIP] Implements different verbosity levels in kubeadm

### PR DESCRIPTION
**Pull request by vbmade2000**
**What this PR does / why we need it**:
This PR adds verbosity level in kubeadm. Currently it is in WIP state.  
**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #340

**Special notes for your reviewer**:
This PR is in WIP state. This is raised just to get approval from you for approach to fix the issue.
**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Comment by vbmade2000** [comment]
Can someone just verify the approach ? PR is in WIP state.  It still requires lot of work to be done. @luxas

File: cmd/kubeadm/app/cmd/init.go [hunk 1]
```diff
@@ -224,13 +233,19 @@ func AddInitOtherFlags(flagSet *flag.FlagSet, cfgPath *string, skipPreFlight, sk
 		criSocket, "cri-socket", "/var/run/dockershim.sock",
 		`Specify the CRI socket to connect to.`,
 	)
+        flagSet.IntVar(
+                verbosityLevel, "verbosity", 0,
+               `Specify the verbosity level of output.`,
+        )
```
Comment (kad) [inline cmd/kubeadm/app/cmd/init.go]: I wouldn't add new flag here. better add one from glog:
https://github.com/spf13/pflag#supporting-go-flags-when-using-pflag

You can look at kube-proxy as example here.

File: cmd/kubeadm/app/cmd/init.go [hunk 2]
```diff
@@ -114,12 +117,18 @@ func NewCmdInit(out io.Writer) *cobra.Command {
...
 
 	cmd := &cobra.Command{
 		Use:   "init",
 		Short: "Run this command in order to set up the Kubernetes master.",
 		Run: func(cmd *cobra.Command, args []string) {
 			var err error
+
+                        // Required for verbose messages to be printed
+                        goflag.Set("logtostderr", "true")
+                        goflag.Set("v", strconv.Itoa(verbosityLevel))
```
Comment (kad) [inline cmd/kubeadm/app/cmd/init.go]: this probably will be not needed once glog flags properly imported into pflag.

**Review by kad — CHANGES_REQUESTED** [review]

**Comment by vbmade2000** [comment]
@kad Thanks for quick suggestions. I'll modify code according to that.

File: cmd/kubeadm/app/cmd/init.go (hunk 2 above)
Comment (vbmade2000) [reply cmd/kubeadm/app/cmd/init.go]: Corrected.

File: cmd/kubeadm/app/cmd/init.go (hunk 1 above)
Comment (vbmade2000) [reply cmd/kubeadm/app/cmd/init.go]: I have confusion with flag.Parse(). kube-proxy doesn't use it currently at least. If I don't use it then pflag doesn't add command line options used by glog package and if I use that like code given below.
```
// NewCmdInit returns "kubeadm init" command.
func NewCmdInit(out io.Writer) *cobra.Command {
	cfg := &kubeadmapiext.MasterConfiguration{}
	legacyscheme.Scheme.Default(cfg)

	var cfgPath string
	var skipPreFlight bool
	var skipTokenPrint bool
	var dryRun bool
	var featureGatesString string
	var criSocket string
	var ignorePreflightErrors []string
        var verbosityLevel int

	cmd := &cobra.Command{
		Use:   "init",
		Short: "Run this command in order to set up the Kubernetes master.",
		Run: func(cmd *cobra.Command, args []string) {
			var err error

                        // Required to include glog package flags with pflag
                	flag.CommandLine.AddGoFlagSet(goflag.CommandLine)

			if cfg.FeatureGates, err = features.NewFeatureGate(&features.InitFeatureGates, featureGatesString); err != nil {
				kubeadmutil.CheckErr(err)
			}

			legacyscheme.Scheme.Default(cfg)
			internalcfg := &kubeadmapi.MasterConfiguration{}
			legacyscheme.Scheme.Convert(cfg, internalcfg, nil)

			ignorePreflightErrorsSet, err := validation.ValidateIgnorePreflightErrors(ignorePreflightErrors, skipPreFlight)
			kubeadmutil.CheckErr(err)

			i, err := NewInit(cfgPath, internalcfg, ignorePreflightErrorsSet, skipTokenPrint, dryRun, criSocket, verbosityLevel)
			kubeadmutil.CheckErr(err)
			kubeadmutil.CheckErr(i.Validate(cmd))
			kubeadmutil.CheckErr(i.Run(out))
		},
	}

	AddInitConfigFlags(cmd.PersistentFlags(), cfg, &featureGatesString)
	AddInitOtherFlags(cmd.PersistentFlags(), &cfgPath, &skipPreFlight, &skipTokenPrint, &dryRun, &criSocket, &ignorePreflightErrors, &verbosityLevel)

        flag.Parse()
	return cmd
}
``` 

It raises following error while building kubeadm.

```
...
...
...
k8s.io/kubernetes/cmd/kubeadm/app/cmd
k8s.io/kubernetes/cmd/kubeadm/app
k8s.io/kubernetes/cmd/kubeadm
go build k8s.io/kubernetes/cmd/kubeadm: /usr/local/go/pkg/tool/linux_amd64/link: signal: killed
!!! [1129 02:30:56] Call tree:
!!! [1129 02:30:56]  1: /home/vbmade2000/go/src/k8s.io/kubernetes/hack/lib/golang.sh:707 kube::golang::build_binaries_for_platform(...)
!!! [1129 02:30:56]  2: hack/make-rules/build.sh:27 kube::golang::build_binaries(...)
!!! [1129 02:30:57] Call tree:
!!! [1129 02:30:57]  1: hack/make-rules/build.sh:27 kube::golang::build_binaries(...)
!!! [1129 02:30:57] Call tree:
!!! [1129 02:30:57]  1: hack/make-rules/build.sh:27 kube::golang::build_binaries(...)
Makefile:92: recipe for target 'all' failed
make: *** [all] Error 1
```

**Comment by kad** [comment]
For current build failures: `cmd/kubeadm/app/cmd/init.go: import of github.com/golang/glog, which is not a direct dependency`  please run `hack/update-bazel.sh` and add changes to `BUILD` files into pull request.

**Comment by vbmade2000** [comment]
@kad Executed hack/update-bazel.sh, build error is resolved.
Can you give some more detail about adding changes to BUILD file ? I am not aware of how build system works using bazel.

**Comment by kad** [comment]
script `hack/update-bazel.sh` will update BUILD files if needed. you just need afterwards do `git status` to see what has changed and do `git add ...` / `git commit ...` as usual for those changed files as well. and then push it again here.

**Comment by vbmade2000** [comment]
Ok Cool.

**Comment by kad** [comment]
when you're ready, please push updated commit.

**Comment by xiangpengzhao** [comment]
@vbmade2000 more info on bazel here:
https://github.com/kubernetes/community/blob/master/contributors/devel/bazel.md

**Comment by vbmade2000** [comment]
@xiangpengzhao Thanks. I'll take a look.

**Comment by vbmade2000** [comment]
@kad On my system commands like ```make check``` etc hangs because of unavailability of enough resources. Also everything is new for me: from Kubernetes to Go so when I get stuck, I push intermediate code. :(

**Comment by vbmade2000** [comment]
I am confused about flag.Parse() call. Despite of calling it, it is not adding glog related command line arguments to kubeadm command line. I used it as mentioned in https://godoc.org/github.com/spf13/pflag#Parse but still it is not working as expected. Moreover I don't see call to Parse() in other kubernetes part like kube=proxy.

**Comment by kad** [comment]
Don't worry, eventually all will be ok :)
For now: please when updating, squash your commits (git rebase interactive and there squash them together). Then, for new changes later on, you'll be able to us "git commit --amend" which will just add new changes to one commit. You can also do "git push -f vbmade2000 ISSUE-340", so you don't need to push merge commits.

For current test failures: when editing, run gofmt on your code. it can be easily done via Go plugins for most of commonly used editors. Or run hack/update-gofmt.sh before committing/pushing new changes.

**Comment by vbmade2000** [comment]
@kad  Yeah. Actually I always run go{fmt,lint,vet} but yesterday after struggling for an hour I found that it's a space and tab issue. I use vim for code and white spaces instead of tab while the existing code seems to use Tab.

**Comment by vbmade2000** [comment]
@kad @luxas Is there any rule like thing to decide verbosity level of message ? If not then on what is the base to decide verbosity level for message ?

**Comment by kad** [comment]
@vbmade2000 I don't think the rule exist yet. However, I think you touched important topic and needed functionality: it will be good to unify all output in all places of kubeadm to use same controllable output mechanism and classify all output to Info/Warnings/Errors.
In my opinion, every warning and error must be shown always. Info type of messages is something that can be controlled based on verbosity flag.

**Comment by vbmade2000** [comment]
@kad glog.Warning() etc functions don't seem to work without --logtostderr=true command line argument . Also pflag.Parse() is required. I am thinking to keep fmt.Println() statements used to print Warnings and Errors, as it is as they print text everytime at least. Also glog doesn't seem to provide log level feature. Any thought ?

**Comment by kad** [comment]
doing `flag.Set("logtostderr", "true")` by default I feel is ok. Parse() also should be ok. 
What do you mean by:
> Also glog doesn't seem to provide log level feature

File: cmd/kubeadm/app/cmd/init.go [hunk 3]
```diff
@@ -114,12 +115,14 @@ func NewCmdInit(out io.Writer) *cobra.Command {
 	var featureGatesString string
 	var criSocket string
 	var ignorePreflightErrors []string
+	var verbosityLevel int
```
Comment (kad) [inline cmd/kubeadm/app/cmd/init.go]: do you really need that variable now ?

Comment (vbmade2000) [reply cmd/kubeadm/app/cmd/init.go]: Nope. I wonder how me and test scripts missed it. I'll remove it.

**Comment by vbmade2000** [comment]
Ok I'll set logtostderr and use Parse(). Regarding `Also glog doesn't seem to provide log level feature`, I was actually thinking of a log level functionality to display log with certain severity level. Example is like log level attribute LOGLEVEL  in Apache Virtual Host if you are aware of that.

**Comment by kad** [comment]
If we are talking about `LogLevel` to limit output of kubeadm with important/tracing messages, my opinion would be that it should be controlled via `verbosity` flag. Default level should include Errors, Warnings and important Info messages. With increased verbosity level, kubeadm can produce more debug and tracing outputs. Is that what you wanted or you have more specific scenarios in mind for Loglevel ?

**Comment by vbmade2000** [comment]
Exactly same thing that I was thinking about. But I agree with the verbosity flag solution you talked about.

**Comment by vbmade2000** [comment]
This PR is messed up due to some git related mistakes. I am going to raise new PR.

**Comment by xiangpengzhao** [comment]
@vbmade2000 okay. I'm going to close the PR. Feel free to create a new PR when you are ready :)

**Comment by luxas** [comment]
@vbmade2000 Let us know when your new PR is ready, no worries :)

**Pull request closed without merging**